velox/functions/prestosql/Zip.cpp

```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <boost/algorithm/string/join.hpp>
+#include "velox/expression/Expr.h"
+#include "velox/expression/VectorFunction.h"
+#include "velox/functions/lib/LambdaFunctionUtil.h"
+
+namespace facebook::velox::functions {
+
+namespace {
+void validateInputTypes(const std::vector<VectorPtr>& inputArgs) {
+  VELOX_USER_CHECK_GT(
+      inputArgs.size(), 1, "zip requires at least two parameters");
+
+  for (auto& arg : inputArgs) {
+    VELOX_USER_CHECK_EQ(
+        arg->type()->kind(),
+        TypeKind::ARRAY,
+        "zip requires arguments of type ARRAY");
+  }
+}
+
+class ZipFunction : public exec::VectorFunction {
+  static const auto kMinArity = 2;
+  static const auto kMaxArity = 7;
+
+ public:
+  /// This class implements the zip function.
+  ///
+  /// DEFINITION:
+  /// zip(ARRAY[T], ARRAY[U]) -> ARRAY(ROW[T,U])
+  /// where we create a ROW[Ti, Ui] for every ith element in ARRAY[T], ARRAY[U].
+  /// The smaller array is padded with nulls.
+  ///
+  /// IMPLEMENTATION:
+  ///  1. The general idea is to create a new dictionary vector for each input
+  ///  array vector and enumerate their indices to create a 1:1 mapping.
+  ///  2. To do this, for each row we determine which is the largest Array
+  ///  and subsequently pad the smaller arrays with nulls.
+  ///  3. Then we take the resultant padded vectors together and create one ROW
+  ///  Vector.
+  ///  4. This forms the base to create the final output Array vector, whose
+  ///  Arrays are the size of the largest input Array.
+  ///
+  ///  Note:
+  ///   - We make no copy's of any constituent elements and are agnostic to
+  ///   types.
+  ///   - For compatibility with Presto a maximum arity of 7 is enforced.
+
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx* context,
+      VectorPtr* result) const override {
+    validateInputTypes(args);
+    const vector_size_t numInputArrays = args.size();
+
+    exec::DecodedArgs decodedArgs(rows, args, context);
+    std::vector<const ArrayVector*> baseVectors(args.size());
```

Comment:
- args.size() -> numInputArrays
- nit: numInputArrays is a bit long; consider shortening to numInputs or numArrays